vhost-device-sound/src/stream.rs

```diff
@@ -239,85 +244,91 @@ impl Default for PcmParams {
     }
 }
 
-pub struct Buffer {
-    data_descriptor: virtio_queue::Descriptor,
+pub struct Request {
     pub pos: usize,
+    len: usize,
     pub message: Arc<IOMessage>,
     direction: Direction,
 }
 
-impl std::fmt::Debug for Buffer {
+impl std::fmt::Debug for Request {
     fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
-        fmt.debug_struct(stringify!(Buffer))
+        fmt.debug_struct(stringify!(Request))
             .field("pos", &self.pos)
+            .field("len", &self.len)
             .field("direction", &self.direction)
             .field("message", &Arc::as_ptr(&self.message))
             .finish()
     }
 }
 
-impl Buffer {
-    pub fn new(
-        data_descriptor: virtio_queue::Descriptor,
-        message: Arc<IOMessage>,
-        direction: Direction,
-    ) -> Self {
+impl Request {
+    pub fn new(len: usize, message: Arc<IOMessage>, direction: Direction) -> Self {
         Self {
             pos: 0,
-            data_descriptor,
+            len,
             message,
             direction,
         }
     }
 
     pub fn read_output(&self, buf: &mut [u8]) -> Result<u32> {
-        let addr = self.data_descriptor.addr();
-        let offset = self.pos as u64;
-        let len = self
-            .message
-            .desc_chain
-            .memory()
-            .read(
-                buf,
-                addr.checked_add(offset)
-                    .expect("invalid guest memory address"),
-            )
-            .map_err(|_| Error::DescriptorReadFailed)?;
-        Ok(len as u32)
+        let mem = self.message.desc_chain.memory();
+
+        let Ok(mut reader) = self.message.desc_chain.clone().reader(mem) else {
+            return Err(Error::DescriptorReadFailed);
+        };
```

Comment:
We use `map_err()` a lot in this crate, so I suggest continuing in that way:

```suggestion
        let mut reader = self
            .message
            .desc_chain
            .clone()
            .reader(mem)
            .map_err(|_| Error::DescriptorReadFailed)?;
```

And in other places.